sdk/go/pulumi/resource.go

```diff
@@ -433,3 +497,92 @@ func Version(o string) ResourceOrInvokeOption {
 		}
 	})
 }
+
+// Converts boxed `Resource`, `ResourceOutput`, or `ResourceInput` to `ResourceInput`.
+func autocastResourceToResourceInput(v interface{}) (ResourceInput, error) {
+	resource, isResource := v.(Resource)
+	if isResource {
+		return NewResourceInput(resource), nil
+	}
+	resourceInput, isResourceInput := v.(ResourceInput)
+	if isResourceInput {
+		return resourceInput, nil
+	}
+	return nil, fmt.Errorf("Expected a Resource, ResourceOutput or a ResourceInput, but got a value of type %v",
+		reflect.TypeOf(v))
+}
+
+// Converts boxed `[]Resource`, `[]ResourceOutput`, `[]ResourceInput` slices to `[]ResourceInput`.
+func autocastSliceToResourceInput(v interface{}) ([]ResourceInput, error) {
```

Comment:
In Go, slices seem to be invariant, one can cast `CustomResource -> interface{} -> Resource` but one cannot cast `[]CustomResource -> interface{} -> []Resource`... Rolling a little helper here. If there is other machinery in the codebase that can do this for our users please LMK.